required/latex-lab/latex-lab-mathcapture.dtx

```diff
@@ -0,0 +1,505 @@
+          { $ }
+      }
+      {
+        \bool_set_true:N \l_@@_collected_bool
+        \@@_grab_inline:w
+      }
+  } % \(
+\cs_gset_protected:Npn \)
+  {
+    \mode_if_math:TF
+      { $ }
+      { \@badmath }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\[, \]}
+%   Again, we need to watch for when \pkg{amsmath} is loaded after this code.
+%   The flag usage here is to cover the case where \cs{[}/\cs{]} is nested
+%   inside another environment.
+%    \begin{macrocode}
+\tl_gput_right:Nn \@kernel@before@begindocument
+  {
+    \cs_gset_protected:Npn \[ % \]
+      {
+        \bool_if:NTF \l_@@_collected_bool
+          { \begin { equation* } }
+          { \@@_grab_eqn:w }
+      } % \[
+    \cs_gset_protected:Npn \]
+      {
+        \bool_if:NTF \l_@@_collected_bool
+          { \end{ equation* } }
+          { \@badmath }
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \begin{macro}{\ensuremath}
+%   A bit of nesting fun to make sure we collect only if required.
+%    \begin{macrocode}
+\cs_gset_protected:Npn \ensuremath #1
+  {
+    \mode_if_math:TF
+      {#1}
+      {
+        \bool_if:NTF \l_@@_collected_bool
+          { \@ensuredmath {#1} }
+          {
+            \bool_set_true:N \l_@@_collected_bool
+            \@@_process:nn { math } {#1}
+            \@ensuredmath {#1}
+            \bool_set_false:N \l_@@_collected_bool
+          }
+      }
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \subsection{\cs{everymath} and \cs{everydisplay}}
+%
+% The business end for grabbing inline math and \enquote{raw} \TeX{}
+% display. Most display math mode is actually handled elsewhere, as we
+% have macro control.
+%    \begin{macrocode}
+\exp_args:No \tex_everymath:D
+  {
+    \tex_the:D \tex_everymath:D
+    \bool_if:NF \l_@@_collected_bool
+      {
+        \bool_set_true:N \l_@@_collected_bool
+        \@@_grab_dollar:w
+      }
+  }
+\exp_args:No \tex_everydisplay:D
+  {
+    \tex_the:D \tex_everydisplay:D
+    \bool_if:NF \l_@@_collected_bool
+      {
+        \bool_set_true:N \l_@@_collected_bool
+        \@@_grab_dollardollar:w
+      }
+  }
+%    \end{macrocode}
+%
+% \subsection{Modifying kernel environments}
+%
+%   We need to cover this even though it is of course not encouraged.
+%    \begin{macrocode}
+\math_register_env:n { eqnarray }
+%    \end{macrocode}
+%
+% Places where math mode is (ab)used.
+%    \begin{macrocode}
+\clist_map_inline:nn
+  { tabular }
+  {
+    \AddToHook{ env / #1 / begin }
+      { \bool_set_true:N \l_@@_collected_bool }
+  }
+%    \end{macrocode}
+%
+% \begin{macro}{\@@_m@th:, \m@th}
+%   Handle non-math use of math mode. At present nesting isn't supported as
+%   \cs{m@th} pops up in a few places that \emph{are} math mode!
+%    \begin{macrocode}
+\cs_new_eq:NN \@@_m@th: \m@th
+\cs_gset_protected:Npn \m@th
+  {
+    \bool_set_true:N \l_@@_collected_bool
+    \@@_m@th:
+  }
+%    \end{macrocode}
+% \end{macro}
+%
+% \subsection{Modifying \pkg{amsmath}}
```

Comment:
Correction, should be: alignat(*)

Reply:
aren't all those automatically processed because they all use the internal amsmath grabbing method into which the code hooks? I thought I had made examples for those too (maybe I haven't).

Reply:
Very likely they are: thus adding them is probably just an "administrative matter": 
adding them to all necessary lists, including registration.

Reply:
I did not find them in any of the current .lvt files.

Reply:
Yes, they are all intended for use as displayed math environments (and some, maybe all, can also be used within a row in gather):

`flalign` is simply a left-aligned, rather than centred, align.

`alignat` and the, not well-documented) "x-aligns" are other variants on the details of the horizontal layout.  They all use `\@startalign` (unlike the "aligned" family that are for use only within mathmode in certain displays).

Reply:
The "-ed" variants all check that they are used only within mathmode.